rollupsync/sync.go

```
@@ -0,0 +1,203 @@
+package rollupsync
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cometbft/cometbft/config"
+	"github.com/cometbft/cometbft/libs/log"
+	sm "github.com/cometbft/cometbft/state"
+	"github.com/cometbft/cometbft/store"
+	"github.com/cometbft/cometbft/types"
+
+	"github.com/cosmos/gogoproto/proto"
+
+	cmtproto "github.com/cometbft/cometbft/proto/tendermint/types"
+	"github.com/cometbft/cometbft/proxy"
+	"github.com/cometbft/cometbft/rollupsync/provider"
+	rstypes "github.com/cometbft/cometbft/rollupsync/types"
+)
+
+type RollupSyncer struct {
+	logger log.Logger
+	cfg    config.RollupSyncConfig
+
+	targetBlockHeight uint64
+
+	// immutable
+	state sm.State
+
+	blockExec *sm.BlockExecutor
+	store     *store.BlockStore
+	proxyApp  proxy.AppConns
+
+	batchProvider  rstypes.BatchProvider
+	outputProvider rstypes.OutputProvider
+
+	blockCh chan rstypes.BlockInfo
+}
+
+func NewRollupSyncer(cfg config.RollupSyncConfig, logger log.Logger, state sm.State, blockExec *sm.BlockExecutor, store *store.BlockStore, proxyApp proxy.AppConns) (*RollupSyncer, error) {
+	batchProvider, outputProvider, err := provider.NewProviders(cfg, logger)
+	if err != nil {
+		return nil, err
+	}
+	return &RollupSyncer{
+		logger: logger,
+
+		state:     state,
+		blockExec: blockExec,
+		store:     store,
+		proxyApp:  proxyApp,
+
+		batchProvider:  batchProvider,
+		outputProvider: outputProvider,
+
+		blockCh: make(chan rstypes.BlockInfo, 100),
+	}, nil
+}
+
+func (rs *RollupSyncer) Start(ctx context.Context) (sm.State, error) {
+	targetBlockHeight, err := rs.outputProvider.GetLatestFinalizedBlock(ctx)
+	if err != nil {
+		return sm.State{}, err
+	}
+	rs.targetBlockHeight = targetBlockHeight
+	if rs.state.LastBlockHeight >= int64(targetBlockHeight) {
+		return rs.state, err
+	}
+	start, _ := rs.blockExec.Store().GetRollupSyncL1Block()
```

`GetRollupSyncBatchChainHeight()`?

---

is it okay to ignore error?

---

it is to start from scratch if there is no saved height